### PR TITLE
feat: support truncate long/double to int/float with option

### PR DESCRIPTION
**Pull request by inocsin**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)
Support truncate long/double to int/float with option, add CompileSpec support
Please close the previous issue https://github.com/NVIDIA/TRTorch/pull/322

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by inocsin** [comment]
@peri044 please review this PR

File: core/conversion/var/Var.cpp [hunk 1]
```diff
@@ -94,10 +94,18 @@ nvinfer1::ITensor* Var::ITensorOrFreeze(ConversionCtx* ctx) {
...
   nvinfer1::ITensor* out;
-
+  auto weights = converters::Weights();
   if (isIValue()) {
-    auto weights = converters::Weights(ctx, ptr_.ivalue->toTensor());
-
+    auto tensor = ptr_.ivalue->toTensor();
+    if (tensor.scalar_type() == at::kLong && ctx->settings.truncate_long_and_double) {
+      weights = converters::Weights(ctx, tensor.toType(at::kInt));
+      LOG_WARNING("Truncate kLong to kInt for IValue");
```
Comment (peri044) [inline core/conversion/var/Var.cpp:102]: Warning: Truncating weight (constant in the graph) from kLong to kInt to indicate that only constants are affected.

File: core/conversion/var/Var.cpp [hunk 2]
```diff
@@ -94,10 +94,18 @@ nvinfer1::ITensor* Var::ITensorOrFreeze(ConversionCtx* ctx) {
...
+    } else if (tensor.scalar_type() == at::kDouble && ctx->settings.truncate_long_and_double) {
+      weights = converters::Weights(ctx, tensor.toType(at::kFloat));
+      LOG_WARNING("Truncate kDouble to kFloat for IValue");
```
Comment (peri044) [inline core/conversion/var/Var.cpp:105]: Same here as well w.r.t warning message

**Review by peri044 — CHANGES_REQUESTED** [review]
You have to add this flag on the python api side as well. Here are the locations

1) https://github.com/NVIDIA/TRTorch/blob/master/py/trtorch/csrc/trtorch_py.cpp#L234
2) https://github.com/NVIDIA/TRTorch/blob/740eb54055d73b10d216c835419acf6ee2e58d31/py/trtorch/csrc/tensorrt_classes.cpp#L95

3) https://github.com/NVIDIA/TRTorch/blob/740eb54055d73b10d216c835419acf6ee2e58d31/py/trtorch/csrc/tensorrt_classes.h#L90

4) https://github.com/NVIDIA/TRTorch/blob/740eb54055d73b10d216c835419acf6ee2e58d31/py/trtorch/csrc/register_tensorrt_classes.cpp#L39

File: core/conversion/var/Var.cpp [hunk 3]
```diff
@@ -94,10 +94,18 @@ nvinfer1::ITensor* Var::ITensorOrFreeze(ConversionCtx* ctx) {
       "Requested either IValue containing a Tensor, or ITensor, however Var type is " << type_name());
 
   nvinfer1::ITensor* out;
-
+  auto weights = converters::Weights();
   if (isIValue()) {
-    auto weights = converters::Weights(ctx, ptr_.ivalue->toTensor());
-
+    auto tensor = ptr_.ivalue->toTensor();
```
Comment (narendasan) [inline core/conversion/var/Var.cpp:99]: We should also check these cases if the setting is not enable to give users a hint to turn on truncating

File: core/conversion/var/Var.cpp [hunk 4]
```diff
@@ -94,10 +94,20 @@ nvinfer1::ITensor* Var::ITensorOrFreeze(ConversionCtx* ctx) {
...
+    if ((tensor.scalar_type() == at::kLong || tensor.scalar_type() == at::kDouble) && !ctx->settings.truncate_long_and_double) {
+      TRTORCH_CHECK(0, "Unable to freeze tensor of type kLong/kDouble into constant layer, try to compile model with truncate_long_and_double ON");
```
Comment (narendasan) [inline core/conversion/var/Var.cpp:101]: Lets just call the types: Int32, Int64, Float64, Float32 throughout the warning messages

Comment (narendasan) [inline core/conversion/var/Var.cpp:101]: You can use `TRTORCH_THROW_ERROR` here

File: core/conversion/var/Var.cpp [hunk 5]
```diff
@@ -97,12 +97,14 @@ nvinfer1::ITensor* Var::ITensorOrFreeze(ConversionCtx* ctx) {
   auto weights = converters::Weights();
   if (isIValue()) {
     auto tensor = ptr_.ivalue->toTensor();
-    if (tensor.scalar_type() == at::kLong && ctx->settings.truncate_long_and_double) {
+    if ((tensor.scalar_type() == at::kLong || tensor.scalar_type() == at::kDouble) && !ctx->settings.truncate_long_and_double) {
+      TRTORCH_CHECK(0, "Unable to freeze tensor of type kLong/kDouble into constant layer, try to compile model with truncate_long_and_double ON");
+    } else if (tensor.scalar_type() == at::kLong && ctx->settings.truncate_long_and_double) {
       weights = converters::Weights(ctx, tensor.toType(at::kInt));
-      LOG_WARNING("Truncate kLong to kInt for IValue");
+      LOG_WARNING("Warning: Truncating weight (constant in the graph) from kLong to kInt to indicate that only constants are affected.");
```
Comment (peri044) [inline core/conversion/var/Var.cpp:104]: @inocsin You can remove "to indicate that only constants are affected." This was just a message for you

File: core/conversion/var/Var.cpp [hunk 6]
```diff
@@ -94,10 +94,20 @@ nvinfer1::ITensor* Var::ITensorOrFreeze(ConversionCtx* ctx) {
...
-
+    auto tensor = ptr_.ivalue->toTensor();
+    if ((tensor.scalar_type() == at::kLong || tensor.scalar_type() == at::kDouble) && !ctx->settings.truncate_long_and_double) {
+      TRTORCH_THROW_ERROR("Unable to freeze tensor of type Int64/Float64 into constant layer, try to compile model with truncate_long_and_double ON");
+    } else if (tensor.scalar_type() == at::kLong && ctx->settings.truncate_long_and_double) {
+      weights = converters::Weights(ctx, tensor.toType(at::kInt));
+      LOG_WARNING("Warning: Truncating weight (constant in the graph) from Int64 to Int32.");
+    } else if (tensor.scalar_type() == at::kDouble && ctx->settings.truncate_long_and_double) {
+      weights = converters::Weights(ctx, tensor.toType(at::kFloat));
+      LOG_WARNING("Warning: Truncating weight (constant in the graph) from Float64 to Float32.");
```
Comment (narendasan) [inline core/conversion/var/Var.cpp:107]: You dont need to say warning, the logging system will add that for you

**Review by narendasan — COMMENTED** [review]
Mostly looks good to go, one comment and resolve comments

**Pull request merged**